index.ts

```diff
@@ -59,12 +59,41 @@ const plugin: Plugin<ReplicatorMetaInput> = {
         }
 
         if (batch.length > 0) {
-            await fetch(`https://${config.host.replace(/\/$/, "")}/e`, {
+            const batchDescription = `${batch.length} event${batch.length > 1 ? 's' : ''}`
+            await fetch(`https://${config.host.replace(/\/$/, '')}/e`, {
                 method: 'POST',
                 body: JSON.stringify(batch),
                 headers: { 'Content-Type': 'application/json' },
-            })
-            console.log(`Flushing ${batch.length} event${batch.length > 1 ? 's' : ''} to ${config.host}`)
+            }).then(
+                (res) => {
+                    if (res.ok) {
+                        console.log(`Flushed ${batchDescription} to ${config.host}`)
+                    } else if (res.status >= 500) {
+                        // Server error, retry the batch later
+                        console.error('Failed to submit ${batchSize} to ${config.host} due to server error', res)
+                        throw new RetryError(`Server error: ${res.status} ${res.statusText}`)
+                    } else {
+                        // node-fetch handles 300s internaly, so we're left with 400s here: skip the batch and move forward
+                        // We might have old events in ClickHouse that don't pass new stricter checks, don't fail the whole export if that happens
+                        console.warn(
+                            `Skipping ${batchDescription}, rejected by ${config.host}: ${res.status} ${res.statusText}`
+                        )
+                    }
+                },
+                (err) => {
+                    if (err.name === 'AbortError' || err.name === 'FetchError') {
+                        // Network / timeout error, retry the batch later
+                        // See https://github.com/node-fetch/node-fetch/blob/2.x/ERROR-HANDLING.md
+                        console.error(
+                            `Failed to submit ${batchDescription} to ${config.host} due to network error`,
+                            err
+                        )
+                        throw new RetryError(`Target is unreachable: ${(err as Error).message}`)
+                    }
+                    // Other errors are rethrown to stop the export
```

Comment:
Should we log here as well? Or does this already happen via the `throw` anyway?

Reply:
unhandled exceptions are [already logged as error](https://github.com/PostHog/posthog/blob/2c2e9ce48e2b51640ba74f08c1982d7d40028057/plugin-server/src/utils/db/sql.ts#L122-L129) by the plugin-server, but I'll add it here for the additional context (batch size and host)